protocol/decoupled-hardfork-features.md

```diff
@@ -0,0 +1,308 @@
+# Purpose
+
+The purpose of this design document is to propose a mechanism for decoupling
+individual features from their hardforks in the OP Stack. This will enhance
+flexibility in feature activation, enabling granular control for testing,
+custom chain configurations, and phased rollouts while maintaining
+coordination across the superchain. The design aims to improve modularity
+without introducing excessive complexity, ultimately supporting more robust
+development and deployment practices of OP Stack software.
+
+# Background
+
+The OP Stack relies on the `RollupConfig` to specify chain parameters,
+including hardfork activation timestamps (e.g., `canyonTime`, `deltaTime`).
+These configurations are centralized in the [`superchain-registry`][scr]
+repository, as TOML files under `superchain/configs/`. For chains that are
+part of the superchain, activations are inherited from a shared timestamp
+called the `superchain_time` to ensure synchronized upgrades.
+
+Currently, when a hardfork timestamp is reached in OP Stack client software,
+all features within that hardfork (e.g., EIPs, protocol changes) activate
+simultaneously. Bundling features in this way simplifies coordination but
+limits flexibility, such as enabling individual features for testing or
+delaying specific ones without forking the config.
+
+In the past, contributors of the OP Stack have had issues surrounding
+hardfork features. For example, long before the Isthmus hardfork, the
+Operator Fee Feature was [proposed][proposed] and approved. The feature
+was considered to be a part of the upcoming Isthmus hardfork. Rust
+implementations of OP Stack components went ahead and implemented the
+feature. For months, the feature sat idle. After it was eventually merged,
+there were issues in the execution client changes for the feature. Removing
+implementations of the feature was considered, but the effort to remove it
+was not seen as worthwhile. After a while, the feature was finally
+complete. It effectively alone delayed the hardfork by a month or so.
+
+This could have been avoided by being able to implement the feature using
+a feature-specific toggle in the `RollupConfig`. The feature toggle would
+be toggled on in the [`superchain-registry`][scr] once the feature is ready.
+In this way, features could be implemented and left in client software
+without being connected to hardfork. The key part: *no changes would
+need to be made to client software to enable the feature*, only enabling
+the feature inside the [`superchain-registry`].
+
+
+# Problem Statement
+
+Bundling features within hardforks limits flexibility:
+
+- All features in a hardfork activate at once, preventing selective enablement
+  for testing or debugging.
+- Custom chains or testnets cannot easily remap or delay individual features
+  without forking entire client implementations.
```

Comment:
Is this something we actually want to support? While a hard fork is in development it can be useful to be able to run devnets with a subset of features that are ready for that testing, but having custom chains pick and choose features from within hard forks creates an explosion of feature combinations that we would need to test and support forever.

I think we should be very, very careful about opening up that Pandora's box.

Reply:
I don't think feature flags should be reused as configuration options. I think it's tempting because without looking closely they're so similar. But it ends up messy because they serve different purposes.

The reason is that feature flags should be lightweight and easy to add and remove (and should be aggressively removed or they become a mess), but configuration options need versioning and deprecation once customers are using them.

Reply:
Open source code doesn't mean you have to support every crazy thing anyone does. We very much get to choose what we support and if other people fork to add things we chose not to support that's fine, but that's their problem, not ours. If it's an important thing that we think we should support, then we make it a feature - either built-in or by providing explicit extension points to allow others to build it in a supported way.

But we just can't afford to support every fork and every feature anyone ever builds.